File: mobile-app/packages/services/lib/src/location_service.dart
import 'package:geolocator/geolocator.dart';
import 'package:geocoding/geocoding.dart';

import 'api_quota_ledger.dart';

/// Function returning the current position. Overridden in tests.
Future<Position> Function() positionGetter = () => Geolocator.getCurrentPosition(
      desiredAccuracy: LocationAccuracy.medium,
      timeLimit: const Duration(seconds: 5),
    );

/// Function mapping coordinates to an ISO 3166 code. Overridden in tests.
Future<String?> Function(double lat, double lon) isoCodeGetter =
    (lat, lon) async {
  final placemarks = await placemarkFromCoordinates(lat, lon);
  return placemarks.isEmpty ? null : placemarks.first.isoCountryCode;
};

/// S-04 – LocationService
class LocationService {
  final ApiQuotaLedger _ledger = ApiQuotaLedger(1);

  Future<Map<String, dynamic>> resolveCountry() async {
    if (!_ledger.isSafe()) throw Exception('quota exceeded');

    final permission = await Geolocator.checkPermission();
    if (permission == LocationPermission.denied ||
        permission == LocationPermission.deniedForever) {
      final req = await Geolocator.requestPermission();
      if (req == LocationPermission.denied ||
          req == LocationPermission.deniedForever) {
        throw Exception('permission denied');
      }
    }

    final pos = await positionGetter();
    final iso2 = await isoCodeGetter(pos.latitude, pos.longitude);
    if (iso2 == null) throw Exception('country not found');

    _ledger.increment();
    return {'iso2': iso2};
  }
}

File: web-app/src/services/LocationService.ts
/* eslint-env browser */
/* global GeolocationPosition navigator */
import { ApiQuotaLedger } from '@/utils/ApiQuotaLedger';
import {
  CountrySettingRepository,
  type CountrySetting
} from '@/repositories/CountrySettingRepository';

/** Function returning the current position. Replaced in tests. */
export let positionGetter = () =>
  new Promise<GeolocationPosition>((resolve, reject) =>
    navigator.geolocation.getCurrentPosition(resolve, reject, { timeout: 5000 })
  );
export const setPositionGetter = (fn: typeof positionGetter) => {
  positionGetter = fn;
};

/** Function mapping coordinates to an ISO code. Replaced in tests. */
export let isoCodeGetter = async (
  lat: number,
  lon: number
): Promise<string | null> => {
  const resp = await fetch(
    `https://api.bigdatacloud.net/data/reverse-geocode-client?latitude=${lat}&longitude=${lon}&localityLanguage=en`
  );
  if (!resp.ok) return null;
  const json = await resp.json();
  return json.countryCode ?? null;
};
export const setIsoCodeGetter = (fn: typeof isoCodeGetter) => {
  isoCodeGetter = fn;
};

/**
 * Service resolving the user's country via browser geolocation.
 */
export class LocationService {
  private ledger: ApiQuotaLedger;
  private repo: CountrySettingRepository;

  constructor(
    repo: CountrySettingRepository = new CountrySettingRepository(),
    ledger: ApiQuotaLedger = new ApiQuotaLedger(1)
  ) {
    this.repo = repo;
    this.ledger = ledger;
  }

  /**
   * Resolve the current country and persist it via the repository.
   */
  async resolveCountry(): Promise<CountrySetting> {
    if (!this.ledger.isSafe()) throw new Error('quota exceeded');
    const pos = await positionGetter();
    const iso2 = await isoCodeGetter(pos.coords.latitude, pos.coords.longitude);
    if (!iso2) throw new Error('country not found');
    const setting: CountrySetting = {
      iso2,
      lastCurrency: 'USD',
      acquired: new Date().toISOString(),
      method: 'GPS'
    };
    await this.repo.save(setting);
    this.ledger.increment();
    return setting;
  }
}


